### ISSUE TITLE
Multidimensional interpolation on regular grids.

### ISSUE THREAD

**Issue by eisberg** [markrogoyski/math-php#382]
What about implementing multidimensional interpolation on regular grids.

We really need a function similar to the implementation in scipy: **interpn** https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interpn.html

Perhaps this problem can be solved with the existing functionality, please tell me

**Comment by markrogoyski** [comment]
Hi @eisberg,

Thank you for your interest in MathPHP.

Multivariate interpolation would be new functionality. I'll update this issue if there is any progress on implementing this feature.

Thanks again for your feedback and suggested new functionality.
Mark

**Comment by Beakerboy** [comment]
It might be Foo to discuss how the data should be organized. Should it be a matrix, or an array of arrays. Should there be separate arrays for the row and column header values?

>         1    2      3        4
> .5     6   8      10      12
> 1      5    6      7        8
>1.5    4    4.7   5.3     6

**Comment by eisberg** [comment]
I have a working version of the RegularGridInterpolator class, which is a port of the Python algorithm. I can share the code

**Comment by eisberg** [comment]
Example of use:

```php
// prepare test data:
$x = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
$y = [10, 11, 12, 13, 14, 15, 16, 17, 18, 19];
$z = [110, 111, 112, 113, 114, 115, 116, 117, 118, 119];

function value_func_3d($x, $y, $z)
{
    return 2 * $x + 3 * $y - $z;
}

for ($x_i = 0; $x_i < count($x); $x_i++) {
    for ($y_i = 0; $y_i < count($y); $y_i++) {
        for ($z_i = 0; $z_i < count($z); $z_i++) {
            $values[$x_i][$y_i][$z_i] = value_func_3d($x[$x_i], $y[$y_i], $z[$z_i]);
        }
    }
}

$points = [$x, $y, $z];

// interpolate:
$interp = new RegularGridInterpolator($points, $values, 'linear');
$result = $interp([2.21, 12.1, 115.9]);
var_dump($result); // float(-75.18)
$result = $interp([2.21, 12.1, 115.9], 'nearest');
var_dump($result); // int(-76)

```

**Comment by markrogoyski** [comment]
Hi @eisberg,

Thank you again for providing the implementation of regular grid interpolation.

I refactored a few things and made some minor improvements to simplify things. I've also added more test cases. Do you mind taking a look at the updated code before I put it in the next release? Let me know if everything looks OK and if there are any test cases you think are missing.

The updated code is currently in the develop branch:
- RegularGridInterpolator: https://github.com/markrogoyski/math-php/blob/develop/src/NumericalAnalysis/Interpolation/RegularGridInterpolator.php
- Unit tests: https://github.com/markrogoyski/math-php/blob/develop/tests/NumericalAnalysis/Interpolation/RegularGridInterpolatorTest.php

Thanks again!
Mark

**Comment by eisberg** [comment]
Hi @markrogoyski 

Thanks! It looks great!

**Comment by eisberg** [comment]
You might want to divide "linear" and "nearest" into different classes?

**Comment by markrogoyski** [comment]
This has been released in version v1.4.0. Thank you for suggesting this feature and providing the initial implementation @eisberg.
